src/Microsoft.OData.Core/UriParser/CustomUriFunctions.cs

```diff
@@ -142,18 +142,62 @@ public static bool RemoveCustomUriFunction(string functionName)
         #region Internal Methods
 
         /// <summary>
-        /// Returns a list of signatures for a function name.
+        /// Returns a list of name-signature pairs for a function name.
         /// </summary>
-        /// <param name="name">The name of the function to look for.</param>
-        /// <param name="signatures">The list of signatures available for the function name.</param>
+        /// <param name="functionCallToken">The name of the function to look for.</param>
+        /// <param name="nameSignatures">
+        /// Output for the list of signature objects for matched function names, with canonical name of the function;
+        /// null if no matches found.
+        /// </param>
+        /// <param name="enableCaseInsensitive">Whether to perform case-insensitive match for function name.</param>
         /// <returns>true if the function was found, or false otherwise.</returns>
-        internal static bool TryGetCustomFunction(string name, out FunctionSignatureWithReturnType[] signatures)
+        internal static bool TryGetCustomFunction(string functionCallToken, out IList<KeyValuePair<string, FunctionSignatureWithReturnType>> nameSignatures,
+            bool enableCaseInsensitive = false)
         {
-            Debug.Assert(name != null, "name != null");
+            Debug.Assert(functionCallToken != null, "name != null");
 
             lock (Locker)
             {
-                return CustomFunctions.TryGetValue(name, out signatures);
+                IList<KeyValuePair<string, FunctionSignatureWithReturnType>> bufferedKeyValuePairs
+                    = new List<KeyValuePair<string, FunctionSignatureWithReturnType>>();
+
+                // Do case-sensitive search first.
+                if (CustomFunctions.ContainsKey(functionCallToken))
+                {
+                    FunctionSignatureWithReturnType[] signatureGroup = null;
+                    CustomFunctions.TryGetValue(functionCallToken, out signatureGroup);

```

Comment:
Rather than look up functionCallToken twice, why not use boolean returned from TryGetValue, as in:

                // Do case-sensitive search first.
                FunctionSignatureWithReturnType[] signatureGroup = null;
                if (CustomFunctions.TryGetValue(functionCallToken, out signatureGroup))
                {
                       foreach (FunctionSignatureWithReturnType sig in signatureGroup)
                      ...